optoforce/src/optoforce/optoforce.py

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python
+
+# Copyright 2015 Shadow Robot Company Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the Free
+# Software Foundation, either version 2 of the License, or (at your option)
+# any later version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import serial
+import struct
+# import binascii
+import rospy
+from geometry_msgs.msg import WrenchStamped
+
+
+class OptoforceData:
+    def __init__(self):
+        self.status = None
+        self.force = []
+
+
+class OptoforceDriver(object):
+    """
+    Driver for Optoforce sensors
+    """
+    _sensor_frame_length = {'31': 16,
+                            '34': 34,
+                            '64': 22}
+
+    # TODO Use Conversion to Newtons (needs sensitivity report) instead  of this simple scaling
+    _scale = 10000
+
+    def __init__(self):
+        """
+        Initialize OptoforceDriver object
+        """
+        self._serial = serial.Serial('/dev/ttyACM0', 1000000, timeout=0.5)
+        self._sensor_type = '34'
+        self._publishers = []
+        self._wrenches = []
+
+        if self._sensor_type == '34':
+            for i in range(4):  # 4 sensors
+                self._publishers.append(rospy.Publisher("optoforce_" + str(i), WrenchStamped, queue_size=100))
+                wrench = WrenchStamped()
+                wrench.header.frame_id = "optoforce_" + str(i)
+                self._wrenches.append(wrench)
+
+    def config(self):
+        # TODO Send config parameters
+        pass
+
+    def run(self):
+        """
+        Runs the read loop.
+        """
+        while not rospy.is_shutdown():
+            s = self._serial.read(self._sensor_frame_length[self._sensor_type])
+            data = self._decode(s)
+            if data:
+                self._publish(data)
```

Comment:
It's not necessary, as we are setting a Serial timeout of 0.5s. It will block until we receive frame_lenght bytes, or until timeout passes. 

We really could set the timeout to None.